### DISCUSSION TITLE
Issue when I trying to test Health Check on application

### DISCUSSION THREAD

**Discussion by lucas-llb** [stryker-mutator/stryker-net#1999] [Q&A] [answered]
Hi everyone,
I got an issue when I trying to test the Health Check on application.
Even when I test the route and the return message, Stryker still got survived mutants on the Health Check Options.
These images below is about my test and can help us:
repo:
https://github.com/lucas-llb/HealthCheckTest

Codes and images:
Startup(edited to code):

> `app.UseRouting();
> app.UseEndpoints(
>     endpoint =>
>         endpoint.MapHealthChecks(
>             "/health",
>             new HealthCheckOptions
>             {
>                 ResponseWriter = async (context, report) =>
>                 {
>                     var result = JsonConvert.SerializeObject(
>                         new HealthCheckResponse(report.Status.ToString())
>                     );
>                     await context.Response.WriteAsync(result);
>                 }
>             }
>         )
> );`

Test(Edited to code):

> `        [Fact]
>         public async Task Given_HealthCheckConfigured_When_ApplicationOn_Then_ShouldReturnApiHealthStatus200_And_ExpectedResponseBody()
>         {
>             var application = applicationFactory.CreateClient();
>             var response = await application.GetAsync("/health");
>             var responseCode = (int)response.StatusCode;
>             var responseContent = await response.Content.ReadAsStringAsync();
>             responseCode.Should().Be(StatusCodes.Status200OK);
>             responseContent.Should().Contain(APPLICATION_STATUS_RESPONSE);
>         }`
> 

Mutation Report:
![image](https://user-images.githubusercontent.com/68111940/163183501-c532d6d0-af39-49b0-a55b-12d3c5c24472.png)

Anyone can help me with it?
Thanks in advance.

**Comment by rouke-broersma** [comment, marked as answer]
What happens to your unit test when you apply the mutation manually to your code? (make the endpoint string empty)

**Reply by lucas-llb** [reply]
the test fails because i recieve a 404 error from the endpoint

**Reply by dupdob** [reply]
I am trying to reproduce, but I have trouble getting the test to run. Is you report based on the provided repo or is the repo an extraction to reproduce it ?
The test project lists strange dependencies: this is a xUnit tests, but most packages are for MsTest. In any case, Stryker cannot find the test, NCrunch cannot find the test and VS cannot find the test either.
I will fix the dependencies to try and diagnose this

**Reply by dupdob** [reply]
Well, I did get it to work and I do not have the problem you report: the mutants appear as killed (the ones here, there are some other survivors)

**Reply by lucas-llb** [reply]
> I am trying to reproduce, but I have trouble getting the test to run. Is you report based on the provided repo or is the repo an extraction to reproduce it ? The test project lists strange dependencies: this is a xUnit tests, but most packages are for MsTest. In any case, Stryker cannot find the test, NCrunch cannot find the test and VS cannot find the test either. I will fix the dependencies to try and diagnose this

Yes it is an extraction of my principal code. I fixed the dependencies, now the test runs properly.
I still got survived mutants in this part of the code.
![image](https://user-images.githubusercontent.com/68111940/163386136-1ab5f3e9-532b-425c-bc95-74499f595ea0.png)

obs.: There is another multants alive, but the principal case is the UseEndpoints().
obs2: If I removed this part of the code, the test will fail.

**Reply by dupdob** [reply]
Here is what I get:
![image](https://user-images.githubusercontent.com/3131866/163669805-ef9b699d-1160-4ee7-9b13-7c67fb6b42c3.png)
As you see, no illegitimate survivor.
I am working on a diagnostic feature that will help to pinpoint the cause of similar problems, hence I am interested in reproducing it to see if my feature provides the correct explanation; but whatever happens here is definitely not covered by it for now.

Is this result consistant across multiple runs ? it may be a timing issue ?
You need to look at the log to understand what may happen, I am afraid

**Reply by lucas-llb** [reply]
Yes, it's consistant in multiple runs, I already disabled the parallelization in tests but it was not sucessfull.
Could you tell me how you run (command) the tests and the stryker? Could it be a configuration issue?

**Reply by lucas-llb** [reply]
Anyone more can help me? I still having this issue.

**Reply by dupdob** [reply]
Can you post here a zip file with the log files (using the `--log-to-file` on the command line) ? As I do not reproduce your problem, it is very difficult to guess what is going on.

**Reply by lucas-llb** [reply]
[logs.zip](https://github.com/stryker-mutator/stryker-net/files/8574543/logs.zip)
This is the log files. Thanks in advance.

**Reply by dupdob** [reply]
The logs look fine. Are you sure you are opening the correct report ?  
According to the logs you provided, you get exactly the same result as I did, with mutants 11 to 16 being killed, like in my snapshot.
What is the mutant's id in your snapshot ? (it appears at the bottom of the page when you select it)?

**Reply by lucas-llb** [reply]
I sent the reports of a project with only the test that I had the problem, but i think that its not getting survivors on it.
But this logs below is from my original project, where I get the issue, I think will be more helpfull.
[logs.zip](https://github.com/stryker-mutator/stryker-net/files/8583774/logs.zip)

**Reply by dupdob** [reply]
I am sorry, but this is getting very confusing. These logs shows that all mutants are killed except one, which is not related to your initial problem if I read the log properly.
Could you please start over by providing;
1) The report with the problem you face
2) The logs
3) ideally the original project or a stripped down project that does reproduce your original issue.

Short from that, I do not feel I am spending my time productively

**Reply by lucas-llb** [reply]
Sorry about the confusion. Let's try to start again.
Unfortunately i can't share the whole code because it's a closed source but i will share what I can.

Report: 
![image](https://user-images.githubusercontent.com/68111940/165839809-8c23352e-3a61-4feb-b090-3429bc7e328c.png)
and
![image](https://user-images.githubusercontent.com/68111940/165839884-5809ebdb-a16f-4aeb-80c9-aed0c98386cf.png)

Logs:
[logs.zip](https://github.com/stryker-mutator/stryker-net/files/8586154/logs.zip)

Stripped down project (stryker works properly here, the issue only happens in my original project, but it has the test I use):
https://github.com/lucas-llb/HealthCheckTest

**Reply by dupdob** [reply]
Thanks for the details. It would be better if I had access to the original project, but I have a theory and a workaround.

## Evidences
I see there is only one survivor (`app.CustomizeHealthCheckPoint()`), it is tested against 4 tests, none of them matching the one in your example.
`
ProgramTests.Given_ThatTheEFCoreQueryAdaptersWhereRegistered_When_ApplicationStarts_Then_AllQueriesShouldBeResolvedWithoutErrors()
ProgramTests.TestSwaggerDocGeneration(environment: \"Staging\", statusCode: NotFound)
ProgramTests.TestSwaggerDocGeneration(environment: \"Production\", statusCode: NotFound)
ProgramTests.TestSwaggerDocGeneration(environment: \"Development\", statusCode: OK)
ProgramTests.Given_SerilogConfigurated_When_ApplicationStarts_Then_LogsShouldAppear()`

## Analysis
This is probably a limitation of coverage analysis: I do not have a clear idea of when and how a webapp is instantiated.
Your initialization logic happens in the app constructor, and and it seems the app is cached/reused across several tests. Caching object across tests results in incomplete coverage information, as in Stryker is not able to identify all tests that would be impacted by a mutation.

## Fix
To this way, we have yet to find a reliable way around this problem. Stryker detects this kind of pattern when capturing coverage information, but only when caching is performed by the mutated project, not by some external framework.
The upcoming analysis function will offer Stryker comments allowing the user to tag code forcing it to be tested against all tests.

## Workaround
As of now, the only workaround in Stryker is to disable coverage analysis. Can you please try and report if this fixes your problem (or not).
Alternatively, I recommend you not to use `IClassFixture` in xUnit and to use constructor and Dispose function instead [see doc](https://xunit.net/docs/shared-context). Tests may be a bit slower in the hand, but it will ensure tests are properly isolated.

**Reply by lucas-llb** [reply]
Hello everyone.
I tried the suggetion about removing the IClassFixture and using the constructor and dispose. Now its working properly.
Thanks a lot @dupdob for the help!
